kitodo-mediaserver-importer/src/main/java/org/kitodo/mediaserver/importer/control/ImporterFlowControl.java

```diff
@@ -55,111 +60,176 @@ public void setImporterProperties(ImporterProperties importerProperties) {
-            }
-
-            Work workResult = workDataReader.read(workDir);
+            LOGGER.info("Starting import of work " + workDir.getName());
 
-            // * Validate the data and the set of files (see below).
-            //
-            // * Read the work data from the mets-mods file.
-            // TODO replace with work reader
-            Work work = new Work("id", "title");
+            Path tempOldWorkFiles = null;
+            Work presentWork = null;
+            Work newWork;
 
-            // Validate import
             try {
-                importValidation.validate(work, mets);
 
-            } catch (ValidationException vex) {
-                LOGGER.error("Validation of " + workDir.getName() + " was unsuccesful, "
-                        + "moving to error folder. Error: " + vex);
-                importerUtils.moveDir(workDir, new File(importerProperties.getErrorFolderPath()));
-                continue;
+                // Get the mets file
+                File mets = new File(workDir, workDir.getName() + ".xml");
+                if (!mets.exists()) {
+                    throw new ImporterException("Mets file not found, expected at " + mets.getAbsolutePath());
+                }
+
+                // Read the work data from the mets/mods file.
+                newWork = workDataReader.read(mets);
+                newWork.setPath(Paths.get(importerProperties.getWorkFilesPath(), newWork.getId()).toString());
+
+                // Validate import data
+                importValidation.validate(newWork, mets);
+
+                //check that naming of folder and mets.xml concedes with workId, otherwise rename
+                if (!StringUtils.equals(newWork.getId(), workDir.getName())) {
+                    LOGGER.info("Id of work to import: " + newWork.getId() + " is different from the mets file name "
+                            + workDir.getName() + ", renaming.");
+
+                    Files.move(mets.toPath(), Paths.get(mets.getParent(), newWork.getId() + ".xml"));
+                    Path newPath = Paths.get(workDir.getParent(), newWork.getId());
+                    Files.move(workDir.toPath(), newPath);
+                    workDir = newPath.toFile();
+                }
+
+                // Check in the database if this work is already present
+                // and if there are identifiers associated to another work.
+                presentWork = workChecker.check(newWork);
+
+                // If the work is already present, it should be replaced.
+                if (presentWork != null) {
+
+                    LOGGER.info("Work " + newWork.getId() + " already present, replacing");
+
+                    newWork.setEnabled(presentWork.isEnabled());
+
+                    // Files created and cached by the fileserver must be deleted. TODO call action instead?
+                    fileDeleter.delete(Paths.get(fileserverProperties.getCachePath(), newWork.getId()));
```

Comment:
Yes, this is about deleting cache files. I think there is a small possibility the CacheDeleteAction will do some special things in the future that are not done by the FileDeleter. To be save I think we should use the Action.